core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AbstractCompositeTermination.java

```diff
@@ -3,89 +3,110 @@
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.List;
+import java.util.Objects;
 
 import ai.timefold.solver.core.impl.phase.scope.AbstractPhaseScope;
 import ai.timefold.solver.core.impl.phase.scope.AbstractStepScope;
 import ai.timefold.solver.core.impl.solver.scope.SolverScope;
 import ai.timefold.solver.core.impl.solver.thread.ChildThreadType;
 
+import org.jspecify.annotations.NullMarked;
+
 /**
  * Abstract superclass that combines multiple {@link Termination}s.
  *
  * @see AndCompositeTermination
  * @see OrCompositeTermination
  */
-public abstract sealed class AbstractCompositeTermination<Solution_>
-        extends AbstractTermination<Solution_>
+@NullMarked
+abstract sealed class AbstractCompositeTermination<Solution_>
+        extends AbstractUniversalTermination<Solution_>
         permits AndCompositeTermination, OrCompositeTermination {
 
     protected final List<Termination<Solution_>> terminationList;
+    protected final List<PhaseTermination<Solution_>> phaseTerminationList;
+    protected final List<SolverTermination<Solution_>> solverTerminationList;
 
     protected AbstractCompositeTermination(List<Termination<Solution_>> terminationList) {
-        this.terminationList = terminationList;
+        this.terminationList = Objects.requireNonNull(terminationList);
+        this.phaseTerminationList = terminationList.stream()
```

Comment:
What happens to universal terminations? Do they get put in both lists?

Reply:
Yes.